partiql-ast/src/main/java/org/partiql/ast/IntervalQualifier.java

```diff
@@ -0,0 +1,272 @@
+         * @param endFieldFractionalPrecision the fractional precision of the end field
+         */
+        public Range(
+                @NotNull DatetimeField startField,
+                @Nullable Integer startFieldPrecision,
+                @NotNull DatetimeField endField,
+                @Nullable Integer endFieldFractionalPrecision
+        ) throws IllegalArgumentException {
+            checkStartField(startField, startFieldPrecision);
+            checkEndField(endField, endFieldFractionalPrecision);
+            this.startField = startField;
+            this.leadingFieldPrecision = startFieldPrecision;
+            this.endField = endField;
+            this.endFieldFractionalPrecision = endFieldFractionalPrecision;
+        }
+
+        /**
+         * Checks that the start field utilizes a valid datetime field and that the leading field precision is valid.
+         * @param startField the start field of the range
+         * @param startFieldPrecision the precision of the leading field
+         * @throws IllegalArgumentException when the start field is not a valid datetime field or the leading field precision is not valid
+         */
+        private static void checkStartField(
+                @NotNull DatetimeField startField,
+                @Nullable Integer startFieldPrecision
+        ) throws IllegalArgumentException {
+            int startCode = startField.code();
+            if (!ALLOWABLE_START_FIELDS.contains(startCode)) {
+                throw new IllegalArgumentException("Invalid interval start field: " + startField.name());
+            }
+            if (startFieldPrecision != null && startFieldPrecision <= 0) {
+                throw new IllegalArgumentException("Invalid leading field precision" + startFieldPrecision);
+            }
+        }
+
+        /**
+         * Checks that the end field utilizes a valid datetime field and that the fractional seconds precision is valid.
+         * @param endField the end field of the range
+         * @param endFieldPrecision the fractional seconds precision of the end field
+         * @throws IllegalArgumentException when the end field is not a valid datetime field or the fractional seconds precision is not valid
+         */
+        private static void checkEndField(
+                @NotNull DatetimeField endField,
+                @Nullable Integer endFieldPrecision
+        ) throws IllegalArgumentException {
+            int code = endField.code();
+            if (!ALLOWABLE_END_FIELDS.contains(code)) {
+                throw new IllegalArgumentException("Invalid interval end field: " + endField.name());
+            }
+            if (endFieldPrecision != null && code != DatetimeField.SECOND) {
+                throw new IllegalArgumentException("Cannot specify fractional seconds precision for interval end field: " + endField.name());
+            }
+            if (endFieldPrecision != null && endFieldPrecision < 0) {
```

Comment:
The spec says the minimum here is `0` and the maximum here is implementation defined, but I would presume there is a useful maximum we could put in place?